src/shaders/pbr.frag

```diff
@@ -0,0 +1,246 @@
+
+  return dotProd / d;
+}
+
+// Specular G, specular geometric attenuation
+// normal: normal direction
+// light: light source direction
+// view: camera direction, aka light outgoing direction
+float specularGeometricAttenuation(vec3 normal,
+                                   vec3 light,
+                                   vec3 view,
+                                   float roughness) {
+  float n_dot_l = max(dot(normal, light), 0.0);
+  float n_dot_v = max(dot(normal, view), 0.0);
+  float ggx1 = geometrySchlickGGX(n_dot_l, roughness);
+  float ggx2 = geometrySchlickGGX(n_dot_v, roughness);
+
+  return ggx1 * ggx2;
+}
+
+// Specular F, aka Fresnel, use Schlick's approximation
+// F0: specular reflectance at normal incidence
+// view: camera direction, aka light outgoing direction
+// half: half vector of light and view
+vec3 fresnelSchlick(vec3 F0, vec3 view, vec3 half) {
+  float v_dot_h = max(dot(view, half), 0.0);
+  return F0 + (1.0 - F0) * pow(1.0 - v_dot_h, 5.0);
+}
+
+// albedoColor: diffuse color from albedoTexture
+//              or from baseColor in the material;
+// metallic: metalness of the surface, from metallicTexture
+//           or from metallic in the material;
+// roughness: roughness of the surface, from roughnessTexture
+//            or from roughness in the material;
+// normal: normal direction
+// light: light source direction
+// view: camera direction, aka light outgoing direction
+// lightRadiance: the radiance of the light,
+//                which equals to intensity * attenuation
+vec3 microfacetModel(vec3 albedoColor,
+                     float metallic,
+                     float roughness,
+                     vec3 normal,
+                     vec3 light,
+                     vec3 view,
+                     vec3 lightRadiance) {
+  vec3 half = normalize(l + v);
+  // compute F0, specular reflectance at normal incidence
+  // for nonmetal, using constant 0.04
+  vec3 F0 = mix(vec3(0.04), albedoColor, metallic);
+  vec3 Fresnel = fresnelSchlick(F0, view, half);
+
+  // Diffuse BRDF
+  // NOTE: energy conservation requires
+  // diffuse + specular <= 1.0, where specular = Fresnel
+  // Also: result does not need to be scaled by 1/PI
+  // See details:
+  // https://seblagarde.wordpress.com/2012/01/08/pi-or-not-to-pi-in-game-lighting-equation/
+  vec3 diffuse = mix(vec3(1.0) - Fresnel, vec3(0.0), metallic) * albedoColor;
+
+  // Specular BDDF
+  float temp =
+      max(4.0 * max(dot(normal, light), 0.0) * max(dot(normal, view), 0.0),
+          Epsilon);
+  vec3 specular = Fresnel *
+                  specularGeometricAttenuation(normal, light, view, roughness) *
+                  normalDistribution(normal, half, roughness) / temp;
+
+  return (diffuse + specular) * lightRadiance * max(dot(normal, light), 0.0);
+}
+
+void main() {
+#if defined(ALBEDO_TEXTURE)
+  vec4 albedoColor = texture(albedoTexture, texCoord);
+#else
+  vec4 albedoColor = Material.baseColor;
+#endif
```

Comment:
As I said above, to follow glTF spec, this needs to be multiplied together. Same for the others.